### ISSUE TITLE
The concretizer should look into all the repositories for spec matches

### ISSUE THREAD

**Issue by alalazo** [spack/spack#15800]
This issue is following https://github.com/spack/spack/pull/15420#issuecomment-606980300 in particular:

> Each use case regarded as work for the concretizer should be recorded as such: many of these things we'd like to solve could possibly be solved, but if we don't account for it in the design, then most likely it will be difficult to support

### Rationale

The use-case exposed here is that of a Spack instance with:
- At least 2 repositories
- One package that has different recipes in both repositories

The two recipes may differ in variants, versions, dependencies etc. but need of course to refer to the same package (meaning the name of the spec must be the same). To create a similar situation one could for example:
```console
$ spack repo create <mydir> test
$ spack repo add <mydir>
```
and then copy and modify a package, say `zlib`, into the new repository with namespace `test`. Let's assume we delete the `optimize` variant and leave just a single `1.0` version in the `test.zlib` recipe.

In a similar situation the current behavior of Spack is to prefer the first repository in the list regardless of the abstract spec that is provided by the user. See for instance the third and fourth commands below:

<details><summary>Console output</summary>

```console
$ spack --version
0.14.1-451-49aa28790

$ spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@1.0%gcc@9.0.1+pic+shared arch=linux-ubuntu18.04-broadwell

$ spack spec builtin.zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@1.2.11%gcc@9.0.1+optimize+pic+shared arch=linux-ubuntu18.04-broadwell

$ spack spec zlib+optimize
Input spec
--------------------------------
zlib+optimize

Concretized
--------------------------------
==> Error: trying to set variant "optimize" in package "zlib", but the package has no such variant [happened during concretization of zlib+optimize]

$ spack spec zlib@1.2.11
Input spec
--------------------------------
zlib@1.2.11

Concretized
--------------------------------
zlib@1.2.11%gcc@9.0.1+pic+shared arch=linux-ubuntu18.04-broadwell
```
</details>

### Description

What a user might expect instead is that the concretizer can consider various aspects of the spec before selecting its namespace and thus that the output above is instead the following for the third and fourth commands:
```console
$ # Even though the test repository has precedence +optimize can be 
$ # satisfied only by the builtin repository, so choose that
$ spack spec zlib+optimize
Input spec
--------------------------------
zlib+optimize

Concretized
--------------------------------
zlib@1.2.11%gcc@9.0.1+optimize+pic+shared arch=linux-ubuntu18.04-broadwell

$ # Even though the test repository has precedence @1.2.11 can be 
$ # satisfied only by the builtin repository, so choose that
$ spack spec zlib@1.2.11
Input spec
--------------------------------
zlib@1.2.11

Concretized
--------------------------------
zlib@1.2.11%gcc@9.0.1+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
```

### Additional information
The issue was raised in a PR proposing to add a second package repository in Spack, that will thus expose more frequently the behaviors above.

### General information

- [x] I have run `spack --version` and reported the version of Spack
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by alalazo** [comment]
@tgamblin Fyi, in case this is not yet handled by the new concretizer

**Comment by scheibelp** [comment]
It's possible, but requires modifying the rules that are input to the SAT solver (so to be clear, I am almost certain it is not yet handled, but it could be). I'd like to finish discussion in https://github.com/spack/spack/pull/15420#issuecomment-606980300 before we push this as required.

Also to be clear: if splitting packages into legacy/non-legacy is the best design choice, then I think it's worthwhile to update the concretizer to handle that. What I'm trying to do right now is figure out whether that's even the best design decision: I don't want to create more work for the concretizer now only to decide later that it didn't need to happen.

**Comment by alalazo** [comment]
>  I'd like to finish discussion in #15420 (comment) before we push this as required.

@scheibelp Hmmm I don't get your point here. I see this issue as decoupled from #15420. Say we decide not to go with a legacy repository bundled with Spack. Spack still supports adding user defined repositories and the behavior is what is described above, which probably is sub-optimal.

**Comment by scheibelp** [comment]
#15420 is the first time anyone has requested this type of behavior from the PR, as mentioned in https://github.com/spack/spack/pull/15420#issuecomment-607391052

**Comment by alalazo** [comment]
That's true the issue has been raised in #15420, but I don't see the implication. Are you saying the current behavior with additional repositories is correct / optimal if we don't go with a legacy repository bundled with Spack  and would be not correct / sub-optimal otherwise?

**Comment by alalazo** [comment]
Closing as outdated
